mysite/views.py

```diff
@@ -1,3 +1,175 @@
+from mysite.models import Project, ProjectBid
 
-# Create your views here.
+User = get_user_model()
+
+
+def home(request):
+    return render(request, 'mysite/home.html')
+
+
+def register(request):
+    if request.method == 'POST':
+        form = UserRegistrationForm(request.POST)
+        if form.is_valid():
+            user_obj = form.cleaned_data
+            username = user_obj['username']
+            email = user_obj['email']
+            password = user_obj['password']
+            user_role = user_obj['user_role']
+            if not (User.objects.filter(username=username).exists() or User.objects.filter(email=email).exists()):
+                User.objects.create_user(username, email, password)
+                user = authenticate(username = username, password = password)
+                # Need recommendation
+                user.user_role = user_role
+                user.save()
+                login(request, user)
+                return HttpResponseRedirect('/')
+            else:
+                raise forms.ValidationError('Looks like a username with that email or password already exists')
+    else:
+        form = UserRegistrationForm()
+    return render(request, 'mysite/register.html', {'form' : form})
+
+
+def profile_view(request):
+    user = request.user
+    return render(request, 'mysite/profile.html', {"user": user})
+
+
+class EditUserProfileView(UpdateView):
+    model = User
+    form_class = UserProfileForm
+    template_name = 'mysite/edit_profile.html'
+
+    def get_object(self, *args, **kwargs):
+        user = get_object_or_404(User, pk=self.kwargs['pk'])
+
+        # We can also get user object using self.request.user  but that doesnt work
+        # for other models.
+
+        return user
+
+    def get_success_url(self, *args, **kwargs):
+        return reverse("home")
+
+
+def change_password(request):
+
+    if request.method == 'POST':
+        form = PasswordChangeForm(request.user, request.POST)
+        if form.is_valid():
+            user = form.save()
+            update_session_auth_hash(request, user)  # Important!
+            messages.success(request, 'Your password was successfully updated!')
+            return redirect('profile')
+        else:
+            messages.error(request, 'Please correct the error below.')
+    else:
+        form = PasswordChangeForm(request.user)
+    return render(request, 'mysite/change_password.html', {
+        'form': form
+    })
+
+
+def project_list(request, template_name='mysite/project_list.html'):
+    projects = Project.objects.all()
+    data = {'object_list': projects}
+    return render(request, template_name, data)
+
+
+def project_create(request, template_name='mysite/project_form.html'):
+    form = ProjectForm(request.POST or None)
+    if form.is_valid():
+        project = form.save(commit=False)
+        project.created_by = request.user
+        project.status = 'Unassigned'
+        form.save()
```

Comment:
override form save method as you are already calling it here and move related code there. Its a good practice not a must do.

Reply:
Feedback Incorporated.